torch/csrc/generic/Tensor.cpp

```diff
@@ -47,32 +77,61 @@
   THHostStorage_(free)(cpu_storage);                           \
 }
 
-#define COPY_FROM_ARRAY_CUDA_HALF(ELTYPE, ARRAY, STORAGE, SIZE) \
+#define COPY_FROM_HALF_ARRAY_CUDA(ELTYPE, ARRAY, STORAGE, SIZE) \
+{ \
+  ELTYPE *arrdata = (ELTYPE*)PyArray_DATA(ARRAY);               \
+  std::unique_ptr<load_real> data_guard(new load_real[SIZE]);   \
+  load_real *data = data_guard.get();                           \
+  for (size_t i=0; i<SIZE; i++) {                               \
+    data[i] = TH_half2float(arrdata[i]);                        \
+  }                                                             \
+  THHostStorage *cpu_storage =                                  \
+      THHostStorage_(newWithData)(data_guard.get(), SIZE);      \
+  cpu_storage->flag &= ~TH_STORAGE_FREEMEM;                     \
+  THCStorage_(copyCPU)(LIBRARY_STATE STORAGE, cpu_storage);     \
+  THHostStorage_(free)(cpu_storage);                            \
+}
+
+#define COPY_FROM_ARRAY_CUDA_HALF(ELTYPE, ARRAY, STORAGE, SIZE)     \
+{ \
+  ELTYPE *arrdata = (ELTYPE*)PyArray_DATA(ARRAY);                   \
+  std::unique_ptr<load_real> data_guard(new load_real[SIZE]);       \
+  load_real *data = data_guard.get();                               \
+  for (size_t i=0; i<SIZE; i++) {                                   \
+    data[i] = arrdata[i];                                           \
+  }                                                                 \
+  THFloatStorage *cpu_storage =                                     \
+      THFloatStorage_newWithData(data_guard.get(), storage_size);   \
+  cpu_storage->flag &= ~TH_STORAGE_FREEMEM;                         \
+  THCudaHalfStorage_copyFloat(LIBRARY_STATE STORAGE, cpu_storage);  \
+  THFloatStorage_free(cpu_storage);                                 \
+}
+
+#define COPY_FROM_HALF_ARRAY_CUDA_HALF(ELTYPE, ARRAY, STORAGE, SIZE)                        \
 { \
-  ELTYPE *arrdata = (ELTYPE*)PyArray_DATA(ARRAY);                  \
-  std::unique_ptr<load_real> data_guard(new load_real[SIZE]);      \
-  load_real *data = data_guard.get();                              \
-  for (size_t i=0; i<SIZE; i++) {                                  \
-    data[i] = arrdata[i];                                          \
-  }                                                                \
-  THFloatStorage *cpu_storage =                                    \
-      THFloatStorage_newWithData(data_guard.get(), storage_size);  \
-  cpu_storage->flag &= ~TH_STORAGE_FREEMEM;                        \
-  THCudaHalfStorage_copyFloat(LIBRARY_STATE STORAGE, cpu_storage); \
-  THFloatStorage_free(cpu_storage);                                \
+  char *arrdata = (char*)PyArray_DATA(ARRAY);                                               \
+  cudaMemcpy(STORAGE->data, arrdata, SIZE * sizeof(float) / 2, cudaMemcpyHostToDevice);     \
```

Comment:
Hmm actually can you still make this overload use `THCStorage_(copyCPU)`? I'm a bit worried about stream handling. This completely ignores it, and using `THC` functions would make it consistent with all other cases for sure

Reply:
I did this
```C
#define COPY_FROM_HALF_ARRAY_CUDA_HALF(ARRAY, STORAGE, SIZE)            \
{ \
  THHalf *arrdata = (THHalf*)PyArray_DATA(ARRAY);                       \
  std::unique_ptr<THHalf> data_guard(arrdata);                          \
  THHostStorage *cpu_storage =                                          \
      THHostStorage_(newWithData)(data_guard.get(), SIZE);              \
  cpu_storage->flag &= ~TH_STORAGE_FREEMEM;                             \
  THCStorage_(copyCPU)(LIBRARY_STATE STORAGE, cpu_storage);             \
}
```
but it segfaults. What's the proper way of using these functions?

Reply:
It's because you wrapped `arrdata` in a `unique_ptr`, and the destructor of `unique_ptr` frees the pointer when it's destructed. Just use `arrdata` to construct the storage.

Reply:
That was my next thought a minute after I wrote this :-). I'm currently compiling that version. Thx.